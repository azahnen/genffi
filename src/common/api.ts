import * as TJS from "typescript-json-schema";
import * as ts from "typescript";

export type VarType = { name: string; type: string; enum: boolean };

export type FunctionType = {
  name: string;
  params: VarType[];
  returnType: string;
  returnOptional: boolean;
  throws: boolean;
};

export type InterfaceType = {
  name: string;
  singleton: boolean;
  handle: boolean;
  fold: boolean;
  functions: FunctionType[];
};

export type StructType = {
  name: string;
  properties: VarType[];
};

export type ApiType = {
  interfaces: InterfaceType[];
  structs: StructType[];
  enums: StructType[];
};

export const generateApi = (source: string): ApiType => {
  // optionally pass ts compiler options
  const compilerOptions = {
    strictNullChecks: true,
  };

  const program = TJS.getProgramFromFiles([source], compilerOptions, "./");

  return analyze(program as unknown as ts.Program);
};

const analyze = (program: ts.Program): ApiType => {
  const typeChecker = program.getTypeChecker();
  const interfaces: InterfaceType[] = [];
  const structs: StructType[] = [];
  const enums: StructType[] = [];

  function inspect(node: ts.Node, tc: ts.TypeChecker) {
    if (node.kind === ts.SyntaxKind.InterfaceDeclaration) {
      const symbol: ts.Symbol = (<any>node).symbol;
      const nodeType = tc.getTypeAtLocation(node);
      const fullyQualifiedName = tc.getFullyQualifiedName(symbol);
      const typeName = fullyQualifiedName.replace(/".*"\./, "");
      const name = typeName;
      if (name.startsWith("Ts2xApi.")) {
        const functions = nodeType
          .getProperties()
          .map((prop) => {
            const propType = tc.getTypeOfSymbolAtLocation(prop, node);
            const propDocTags = prop.getJsDocTags(tc);
            const propTypeName = tc.typeToString(propType);
            //console.log(name, prop.getName(), propTypeName);

            const parsed = parseFunction(
              prop.getName(),
              propTypeName,
              propDocTags,
              enums
            );
            //console.log(parsed);
            //console.log(toC(parsed!));

            return parsed;
          })
          .filter((f) => f !== null);

        const docTags = symbol.getJsDocTags(tc);

        interfaces.push({
          name: name.replace("Ts2xApi.", ""),
          singleton: docTags.some((tag) => tag.name === "singleton"),
          handle: docTags.some((tag) => tag.name === "handle"),
          fold: docTags.some((tag) => tag.name === "fold"),
          functions: functions as FunctionType[],
        });
      }
    } else if (node.kind === ts.SyntaxKind.TypeAliasDeclaration) {
      const symbol: ts.Symbol = (<any>node).symbol;
      const nodeType = tc.getTypeAtLocation(node);
      const fullyQualifiedName = tc.getFullyQualifiedName(symbol);
      const typeName = fullyQualifiedName.replace(/".*"\./, "");
      const name = typeName;
      if (name.startsWith("Ts2xApi.")) {
        const props = nodeType
          .getProperties()
          .map((prop) => {
            const propType = tc.getTypeOfSymbolAtLocation(prop, node);
            const propTypeName = tc.typeToString(propType);

            return {
              name: prop.getName(),
              type: propTypeName,
              //TODO: multiple inspect runs, enums first
              enum: enums.find((e) => e.name === propTypeName) !== undefined,
            };
          })
          .filter((f) => f !== null);

        structs.push({
          name: name.replace("Ts2xApi.", ""),
          properties: props,
        });
      }
    } else if (node.kind === ts.SyntaxKind.EnumDeclaration) {
      const symbol: ts.Symbol = (<any>node).symbol;
      const fullyQualifiedName = tc.getFullyQualifiedName(symbol);
      const typeName = fullyQualifiedName.replace(/".*"\./, "");
      const name = typeName;
      if (name.startsWith("Ts2xApi.")) {
        const props: VarType[] = [];

        for (const [key, value] of symbol.exports?.entries() || []) {
          const propDecl = value.valueDeclaration?.getText() || "";
          const propType =
            propDecl.includes("=") && propDecl.includes('"')
              ? "string"
              : "number";
          props.push({ name: key as string, type: propType, enum: false });
        }

        enums.push({
          name: name.replace("Ts2xApi.", ""),
          properties: props,
        });
      }
    } else {
      ts.forEachChild(node, (n) => inspect(n, tc));
    }
  }

  program.getSourceFiles().forEach((sourceFile, _sourceFileIdx) => {
    inspect(sourceFile, typeChecker);
  });

  return { interfaces, structs, enums };
};

const toC = (f: FunctionType): string => {
  const params = f.params
    .map((p) => `const ${toCType(p.type)} ${p.name}`)
    .join(", ");

  return `${toCType(f.returnType)} ${f.name}(${params});`;
};

const toCType = (type: string): string => {
  if (type === "boolean") {
    return "bool";
  }
  if (type === "string") {
    return "char*";
  }
  if (type === "bigint") {
    return "long long int";
  }
  if (type === "Uint8Array") {
    return "char*";
  }
  throw new Error(`Unknown type: ${type}`);
};

const toJavaType = (type: string): string => {
  if (type === "boolean") {
    return "boolean";
  }
  if (type === "string") {
    return "String";
  }
  if (type === "bigint") {
    return "long";
  }
  if (type === "Uint8Array") {
    return "byte[]";
  }
  throw new Error(`Unknown type: ${type}`);
};

const parseFunction = (
  name: string,
  body: string,
  docTags: ts.JSDocTagInfo[],
  enums: StructType[]
): FunctionType | null => {
  const functionRegex = /\(([^)]*)\)\s*=>\s*([^\s]+)/;
  const match = body.match(functionRegex);
  if (match) {
    const params =
      match[1].length === 0
        ? []
        : match[1].split(",").map((param) => {
            const [name, type] = param
              .trim()
              .split(":")
              .map((p) => p.trim());
            return {
              name,
              type,
              //TODO: multiple inspect runs, enums first
              enum: enums.find((e) => e.name === name) !== undefined,
            };
          });
    const returnType = match[2];

    const throws = docTags.some((tag) => tag.name === "throws");
    const returnOptional = docTags.some((tag) => tag.name === "optional");

    return { name, params, returnType, returnOptional, throws };
  }
  return null;
};
