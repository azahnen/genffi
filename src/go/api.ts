import {
  ApiType,
  FunctionType,
  InterfaceType,
  StructType,
  VarType,
} from "../common/api.ts";
import { Generator } from "../common/index.ts";
import { Result } from "../common/io.ts";
import { firstLetterUpperCase } from "../common/schema.ts";
import { go } from "./lang.ts";
import { funcHeader, generateFile } from "./shared.ts";

export const generateGoApi = (
  name: string,
  api: ApiType,
  pkg: string,
  dataNs: string[],
  filePrefixes: Record<string, string> = {}
) => {
  const result: Result = { name, files: [] };

  result.files.push(
    generateFile("gen", "", pkg, filePrefixes, generateApi(api))
  );

  return result;
};

const generateApi =
  (api: ApiType): Generator =>
  (name: string, pkg: string): string => {
    let code = `package ${pkg}  

${api.enums.map(enumToGo).join("\n")}

${api.structs.map(structToGo).join("\n")}

${api.interfaces.map(intfaceToGo).join("\n")}
      `;

    return code;
  };

const intfaceToGo = (i: InterfaceType): string => {
  const funcs = i.functions.map(funcHeader).join("\n  ");

  return `
type ${i.name} interface {
  ${funcs}
}
`;
};

const structToGo = (i: StructType): string => {
  const props = i.properties.map(propToGo).join("\n  ");

  return `
type ${i.name} struct {
  ${props}
}
  `;
};

const enumToGo = (s: StructType): string => {
  const props = s.properties
    .map(
      (p, i) =>
        `${s.name}${p.name} ${s.name} = ${
          p.type === "string" ? `"${p.name}"` : i
        }`
    )
    .join("\n  ");

  return `
type ${s.name} string
	
const (
	${props}
)
		`;
};

const propToGo = (v: VarType): string => {
  return `${firstLetterUpperCase(v.name)} ${go.type(v.type)}`;
};
