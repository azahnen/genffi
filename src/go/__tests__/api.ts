import { describe, it, expect } from "@jest/globals";

import { generateGoApi } from "../api.ts";
import { fixtures } from "../../__tests__/fixtures.ts";

describe("generateGoApi", () => {
  for (const fixture of fixtures) {
    it(fixture.name, () => {
      const result = generateGoApi(fixture.name, fixture.api, "test", []);

      expect(result).toMatchSnapshot();
    });
  }
});
