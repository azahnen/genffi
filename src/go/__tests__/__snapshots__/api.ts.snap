// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateGoApi handle and fold 1`] = `
{
  "files": [
    {
      "content": "package test  


type SourceType string
	
const (
	SourceTypeFS SourceType = "FS"
  SourceTypeGIT SourceType = "GIT"
  SourceTypeS3 SourceType = "S3"
  SourceTypeHTTP SourceType = "HTTP"
)
		

type ContentType string
	
const (
	ContentTypeBLOBS ContentType = "BLOBS"
  ContentTypeCONFIGS ContentType = "CONFIGS"
)
		


type StoreCfg struct {
  Sources []StoreSource
}
  

type StoreSource struct {
  Src string
  Typ SourceType
  Content ContentType
}
  

type StoreStatus struct {
  Ready bool
  Progress float64
}
  


type Stores interface {
  Open(cfg StoreCfg) Store
  Close(store Store) 
}


type Store interface {
  Config() StoreCfg
  Status() StoreStatus
  BlobReader(basePath string) BlobReader
}


type BlobReader interface {
  Path() string
  ReadOnlyBlob(path string) ReadOnlyBlob
  SyncBlobsToFs(path string) (string, error)
}


type ReadOnlyBlob interface {
  Path() string
  Exists() bool
  Size() (int64, bool)
}

      ",
      "path": "test/gen.go",
    },
  ],
  "name": "handle and fold",
}
`;

exports[`generateGoApi simple command handler 1`] = `
{
  "files": [
    {
      "content": "package test  






type CommandHandler interface {
  Execute(command []byte) []byte
}

      ",
      "path": "test/gen.go",
    },
  ],
  "name": "simple command handler",
}
`;
